code/__HELPERS/sorts/__main.dm

```diff
@@ -0,0 +1,647 @@
+	//These are macros used to reduce on proc calls
+#define fetchElement(L, i) (associative) ? L[L[i]] : L[i]
+
+	//Minimum sized sequence that will be merged. Anything smaller than this will use binary-insertion sort.
+	//Should be a power of 2
+#define MIN_MERGE 32
+
+	//When we get into galloping mode, we stay there until both runs win less often than MIN_GALLOP consecutive times.
+#define MIN_GALLOP 7
+
+//This is a global instance to allow much of this code to be reused. The interfaces are kept separately
+GLOBAL_DATUM_INIT(sortInstance, /datum/sortInstance, new())
+/datum/sortInstance
+	//The array being sorted.
+	var/list/L
+
+	//The comparator proc-reference
+	var/cmp = /proc/cmp_numeric_asc
+
+	//whether we are sorting list keys (0: L[i]) or associated values (1: L[L[i]])
+	var/associative = 0
+
+	//This controls when we get *into* galloping mode.  It is initialized	to MIN_GALLOP.
+	//The mergeLo and mergeHi methods nudge it higher for random data, and lower for highly structured data.
+	var/minGallop = MIN_GALLOP
+
+	//Stores information regarding runs yet to be merged.
+	//Run i starts at runBase[i] and extends for runLen[i] elements.
+	//runBase[i] + runLen[i] == runBase[i+1]
+	var/list/runBases = list()
+	var/list/runLens = list()
+
+
+/datum/sortInstance/proc/timSort(start, end)
```

Comment:
This is still relatively pathed on TG. How dare they

Reply:
Nobody wants to touch this code for some reason. It's all awkwardly designed as a singleton